### ISSUE TITLE
オプティカルフロー処理にかけたい画像を正規化処理する. 

### ISSUE THREAD

**Issue by Anna-Oura** [kikuchiken-waseda/OuraAnna#67]
- [x] 教師データ作成の自動化. 
- [x] 正規化処理の自動化. 
- [ ] 正規化処理を行う.

**Comment by Anna-Oura** [comment]
# 参考
- https://github.com/kikuchiken-waseda/OuraAnna/issues/55
- https://github.com/kikuchiken-waseda/rtMRI-semantic-segmentation/tree/main/pca_CharacteristicPoint

**Comment by Anna-Oura** [comment]
# アノテーション方法
labelme を用いた方法. 毎度忘れる. 
## 手順
- `$ conda activate mask_rcnn` で仮想環境 "mask_rcnn" に入る. 
- `$ labelme` を実行してアプリを立ち上げる. 
- `$ conda deactivate` で仮想環境から出る.

**Comment by Anna-Oura** [comment]
# 教師データ作成までの処理
## 手順
- json と png が保存されているファイルがある状況で以下のコードを実行. 
```
$ python3 runner.py
```
- これを実行すると, `seg-data` というファイルが作成され, その中に images と mask が保管される. 
- 次に, 以下のコードを実行. 
```
$ python3 teaching_data.py
```
- これを実行すると, `result` のファイル内に音素＆ファイル毎で以下のような教師データが保存される. 
![img_001](https://user-images.githubusercontent.com/61837100/134294478-968878c0-fac6-475b-b5ab-1224588fe19a.png)
## プログラム
音素名やファイル名は適宜変更する. 
### runner.py
```python3
import json
import os
import glob
import shutil

# 画像関係
import numpy as np
import cv2
from PIL import Image

# 画像表示
import matplotlib.pyplot as plt
IMAGE_SIZE = 256

# データにかけたいファイルを指定
name = ["MU1-1-20171110-4",	"MU1-16-20190111-4", "MU1-5-20171225-4", "MU1-11-20180305-4", "MU1-17-20190111-4", "MU1-7-20171110-4", "MU1-14-20180808-4", "MU1-2-20171225-4", "MU1-9-20171225-4"]
onso = "o"

for num in range(0, len(name)):
    # データのリスト
    json_list = glob.glob('./img/vowel/' +onso+ '/' +str(name[num])+ '/*.json')
    img_list = [f.replace('json', 'png') for f in json_list]
    #print(len(json_list))

    no = 1

    # アノテーションデータ読み込み
    with open(json_list[no]) as f:
        data = json.loads(f.read())

    # 1つだけ取り出す
    shape = data['shapes'][0]
    label = shape['label']
    points = shape['points']
    shape_type = shape['shape_type']
    print('[label]', label)
    print('[shape_type]', shape_type)
    print('[points]', points)

    # 画像読み込み
    img = cv2.imread(img_list[no])
    #img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

    # アノテーション部分
    mask = np.zeros((img.shape[0], img.shape[1]), dtype=np.uint8)
    mask = cv2.fillPoly(mask, np.int32([points]), 1)

    # 横並びに表示
    fig = plt.figure(figsize=(12, 6))
    ax1 = fig.add_subplot(1, 2, 1)
    ax2 = fig.add_subplot(1, 2, 2)
    ax1.imshow(img)
    ax2.imshow(mask, cmap='gray')

    # フォルダ作成
    data_dir = 'seg-data'
    #val_dir = 'val'
    if not os.path.exists(data_dir):
        os.mkdir(data_dir)
        os.mkdir(data_dir + '/' +onso)
        os.mkdir(data_dir + '/' +onso+ '/images')
        os.mkdir(data_dir + '/' +onso+ '/masks')

    # 画像一枚一枚に対して処理
    for ind, file in enumerate(json_list):
        points = []
        with open(file) as f:
            data = json.loads(f.read())
            for s in data['shapes']:
                points.append(s['points'])

        if points:
            # 画像データを読み込み画像サイズ取得
            img_path = file.replace('json', 'png')
            img = cv2.imread(img_path)

            # ファイル名
            file_name = os.path.basename(img_path)

            # 画像保存用手順
            s1 = file_name.replace('.png', '') # 画像の順番ごとに
            os.makedirs(data_dir + '/' +onso+ '/images/' +name[num] , exist_ok=True)
            os.makedirs(data_dir + '/' +onso+ '/images/' +name[num]+ '/' +s1, exist_ok=True)
            os.makedirs(data_dir + '/' +onso+ '/masks/' +name[num] , exist_ok=True)
            os.makedirs(data_dir + '/' +onso+ '/masks/' +name[num]+ '/' +s1, exist_ok=True)

            # jsonのアノテーションデータ
            mask = np.zeros((img.shape[0], img.shape[1]), dtype=np.uint8)
            for p in points:
                mask = cv2.fillPoly(mask, np.int32([p]), 1)

            # リサイズ
            img = cv2.resize(img, (IMAGE_SIZE, IMAGE_SIZE), interpolation=cv2.INTER_NEAREST)
            mask = cv2.resize(mask, (IMAGE_SIZE, IMAGE_SIZE), interpolation=cv2.INTER_NEAREST)

            # 保存
            #file_name = file_name.replace('jpg', 'png')
            maskim = Image.fromarray(np.uint8(mask))
            maskim.save(f'seg-data/' +onso+ '/masks/' +name[num]+ '/' +s1+ '/' +s1+ '.png')
            cv2.imwrite(f'seg-data/' +onso+ '/images/' +name[num]+ '/' +s1+ '/' +s1+ '.png' , img)
```
### teaching_data.py
```python3
# -*- coding: utf-8 -*-

import glob
import cv2
import numpy as np
import os

import torch
from torch.utils.data import DataLoader
from torch.utils.data import Dataset as BaseDataset
#import torch.nn as nn
#import torch.nn.functional as F
#import segmentation_models_pytorch as smp

import matplotlib.pyplot as plt

def visualize(**images):
    """PLot images in one row."""
    n = len(images)
    plt.figure(figsize=(3, 3))
    for i, (name, image) in enumerate(images.items()):
        plt.subplot(1, n, i + 1)
        plt.xticks([])
        plt.yticks([])
        #plt.title(' '.join(name.split('_')).title())
        plt.imshow(image)
    #plt.show()
    return plt

# データセット
class Dataset(BaseDataset):
    CLASSES = ['background', 'koukuu']
    def __init__(
            self,
            images_dir,
            masks_dir,
            classes=None,
            augmentation=None,
            preprocessing=None,
    ):
        self.ids = os.listdir(images_dir)
        self.images_fps = [os.path.join(images_dir, image_id) for image_id in self.ids]
        self.masks_fps = [os.path.join(masks_dir, image_id) for image_id in self.ids]

        # convert str names to class values on masks
        self.class_values = [classes.index(cls.lower()) for cls in classes]

        self.augmentation = augmentation
        self.preprocessing = preprocessing

    def __getitem__(self, i):

        # read data
        image = cv2.imread(self.images_fps[i])
        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

        # mask
        mask = cv2.imread(self.masks_fps[i], 0)
        masks = [(mask == v) for v in self.class_values]
        mask = np.stack(masks, axis=-1).astype('float')

        # apply augmentations
        if self.augmentation:
            sample = self.augmentation(image=image, mask=mask)
            image, mask = sample['image'], sample['mask']

        # apply preprocessing
        if self.preprocessing:
            sample = self.preprocessing(image=image, mask=mask)
            image, mask = sample['image'], sample['mask']

        return image, mask

    def __len__(self):
        return len(self.ids)

# データにかけたいファイルを指定
name = ["MU1-1-20171110-4",	"MU1-16-20190111-4", "MU1-5-20171225-4", "MU1-11-20180305-4", "MU1-17-20190111-4", "MU1-7-20171110-4", "MU1-14-20180808-4", "MU1-2-20171225-4", "MU1-9-20171225-4"]
onso = "o"

# フォルダ作成
result_dir = 'result'
if not os.path.exists(result_dir):
    os.mkdir(result_dir)
    os.mkdir(result_dir + '/' +onso )

for num in range(0, len(name)):
    # 処理回数カウント # 適宜変更
    count = len(glob.glob('./img/vowel/' +onso+ '/' +str(name[num])+ '/*.png'))
    #print(count)

    for j in range(1, count+1):

        # 画像読み込み
        data_dir = './seg-data/' +onso+ '/images/' +str(name[num])+ '/img_00' +str(j)
        mask_dir = './seg-data/' +onso+ '/masks/' +str(name[num])+ '/img_00' +str(j)

        if count <10:
            data_files = glob.glob(os.path.join(data_dir, "*"))
        else:
            data_files = glob.glob(os.path.join(data_dir, "*"))

        f = str(data_files[0])
        image_src = cv2.imread(f)
        image_src = cv2.cvtColor(image_src, cv2.COLOR_BGR2RGB)

        # 画像保存用手順
        s1 = f[-11:]
        s2 = s1.replace('.png', '')
        #print(s2, s3)
        os.makedirs(result_dir + '/' +onso+ '/' +name[num] , exist_ok=True)

        # データ確認
        dataset = Dataset(os.path.join(data_dir), os.path.join(mask_dir), classes=['koukuu'])

        image, mask = dataset[0] # get some sample
        im1 = visualize(
            #image=image,
            koukuu=mask.squeeze()
        )

        im1.savefig('./result/' +onso+ '/' +str(name[num])+ '/' + s2 + '.png')
```

**Comment by Anna-Oura** [comment]
# 主成分分析処理
一枚一枚やるのは面倒すぎるので音素毎で一気にできるようにしました. 
処理時間はそんなにかからなさそう. 
## 手順
- 上記の `教師データ作成までの処理` まで実行し, result に教師データである五角形をプロットした図が保存されている状態. 
- 以下のコードを実行. 
```
$ python3 PCA.py
```
- うまくいけば, 新しく result_pca というフォルダが作成され, その中に回転処理後MRIが保管されているはず. 
## プログラム
onso 部分は適宜変更. 
```python3
import cv2
import numpy as np
import math
import glob
import os

def drawAxisArrow(img, ce, prmpt, colour, scale):
    # 角度
    angle = math.atan2(ce[1] - prmpt[1], ce[0] - prmpt[0])
    hypotenuse = math.sqrt((ce[1] - prmpt[1]) ** 2 + (ce[0] - prmpt[0]) ** 2)

    # スケール変数により矢印長を決定
    q = (ce[0] - scale * hypotenuse * math.cos(angle), ce[1] - scale * hypotenuse * math.sin(angle) )

    # 矢印を描画
    cv2.arrowedLine(img, (int(ce[0]), int(ce[1])), (int(q[0]), int(q[1])), colour, thickness=2, tipLength=0.3)

def OrientationPCA(lbno, pts, img):
    sz = len(pts[0])
    data_pts = np.empty((sz, 2), dtype=np.float64)
    data_pts[:,0] = pts[1][:]
    data_pts[:,1] = pts[0][:]

    # 主成分分析
    #
    # mean : 投影前に差し引かれ、逆投影後に追加された平均値
    # eigenvalues : 共変動行列の固有値
    # eigenvectors : 共変動行列の固有ベクトル
    mean = np.empty((0))
    mean, eigenvectors, eigenvalues = cv2.PCACompute2(data_pts, mean)

    # 平均
    center = (mean[0,0], mean[0,1])
    cv2.circle(img, (int(center[0]), int(center[1])), 3, (255, 0, 255), 2)

    sc = 0.50

    # 第1主成分軸 角度計算
    angle = math.atan2(eigenvectors[0,1], eigenvectors[0,0])

    # 角度描画
    deg = -1 * angle * 180.0 / math.pi
    degtext = "L%d %.1f deg" % (lbno, deg)
    dx = int(center[0])
    dy = int(center[1])
    #cv2.putText(img, degtext, (dx,dy), cv2.FONT_HERSHEY_SIMPLEX, 0.4, (255,0,255),1,cv2.LINE_AA)

    # 寄与率
    v1 = eigenvalues[0,0] / eigenvalues.sum()
    v2 = eigenvalues[1,0] / eigenvalues.sum()
    degtext = "Cont. rate %.2f %.2f" % (v1, v2)
    #cv2.putText(img, degtext, (dx,dy), cv2.FONT_HERSHEY_SIMPLEX, 0.4, (255,0,255),1,cv2.LINE_AA)

    result = []
    # 第1主成分軸
    p1 = (center[0] + sc * eigenvectors[0,0] * eigenvalues[0,0], center[1] + sc * eigenvectors[0,1] * eigenvalues[0,0])
    length1 = (sc * eigenvectors[0,0] * eigenvalues[0,0], sc * eigenvectors[0,1] * eigenvalues[0,0]) #それぞれの矢印のx方向幅とy方向幅を算出

    # 第2主成分軸
    p2 = (center[0] - sc * eigenvectors[1,0] * eigenvalues[1,0], center[1] - sc * eigenvectors[1,1] * eigenvalues[1,0])
    length2 = (sc * eigenvectors[1,0] * eigenvalues[1,0], sc * eigenvectors[1,1] * eigenvalues[1,0])

    return deg, dx, dy, length1, length2, v1, p1

# データにかけたいファイルを指定
name = ["MU1-1-20171110-4",	"MU1-16-20190111-4", "MU1-5-20171225-4", "MU1-11-20180305-4", "MU1-17-20190111-4", "MU1-7-20171110-4", "MU1-14-20180808-4", "MU1-2-20171225-4", "MU1-9-20171225-4"]
onso = "o"

# フォルダ作成
result_dir = 'result_pca'
if not os.path.exists(result_dir):
    os.mkdir(result_dir)
    os.mkdir(result_dir + '/' +onso )

for num in range(0, len(name)):
    # 画像データ読込
    list1 = glob.glob('./result/' +onso+ '/' +str(name[num])+ '/*.png') #PCA分析用画像
    list2 = glob.glob('./img/vowel/' +onso+ '/' +str(name[num])+ '/*.png') #回転用画像

    os.makedirs(result_dir + '/' +onso+ '/' +str(name[num]), exist_ok=True)
    os.makedirs(result_dir + '/' +onso+ '/' +str(name[num])+ '/kaiten', exist_ok=True)
    os.makedirs(result_dir + '/' +onso+ '/' +str(name[num])+ '/syuseibun', exist_ok=True)

    for j in range(0, len(list1)):

        print(str(j) +'回目の処理')
        print(list1[j])
        #print(list2[j])

        src = cv2.imread(str(list1[j]), cv2.IMREAD_COLOR) #PCA分析用画像
        src2 = cv2.imread(str(list2[j])) #回転用画像
        if src is None:
            print('Could not open or find the image: ', args.input)
            exit(0)

        # 結果表示用 画像の脱色処理
        #result = (src // 2) + 127

        # 画像のサイズの変更
        src = cv2.resize(src, dsize=(256,256))

        # グレイスケール,二値化
        gray = cv2.cvtColor(src, cv2.COLOR_BGR2GRAY)
        _, bw = cv2.threshold(gray, 0, 255, cv2.THRESH_BINARY | cv2.THRESH_OTSU)
        bw = cv2.bitwise_not(bw)

        # ラベリング実行
        nLabels, labelImages = cv2.connectedComponents(bw)

        p1_length = []
        p2_length = []
        info = []
        for i in range(1, nLabels):
            # 対象オブジェクト座標の取得
            ls = np.where(labelImages == i)

            # PCA 第1,2主成分軸 処理
            kakudo, X, Y, long1, long2, v1, p1 = OrientationPCA(i, ls, bw)
            #print(X)
            l1 = list(long1)
            l2 = [kakudo, X, Y, v1, p1]
            l3 = list(long2)
            #print(math.sqrt(l1[0]**2 + l1[1]**2))
            p1_length.append(math.sqrt(l1[0]**2 + l1[1]**2)) #第１主成分の長さ
            p2_length.append(math.sqrt(l3[0]**2 + l3[1]**2)) #第２主成分の長さ
            info.append(l2)

        # 主成分軸の長さで場合分けして使う情報を選択
        use_info = info[p1_length.index(sorted(p1_length)[0])] #角度などの情報を取得
        use_p1_length = p1_length[p1_length.index(sorted(p1_length)[0])] #第１主成分の長さを取得
        use_p2_length = p2_length[p1_length.index(sorted(p1_length)[0])] #第２主成分の長さを取得

        #print('p1の長さ')
        #print(use_p1_length)

        #print(use_info)
        Xdesu = use_info[1]
        Ydesu = use_info[2]
        Kakudodesu = use_info[0]
        drawAxisArrow(src, (Xdesu, Ydesu), use_info[4], (0, 0, 255), use_info[3])

        # 画像の中心をオブジェクトの重心に
        M = np.float32([[1, 0, 128-Xdesu], [0, 1, 128-Ydesu]])
        shifted_img = cv2.warpAffine(src2, M, (256, 256))

        #スケールを指定
        scale = 1.0

        #getRotationMatrix2D関数を使用
        if kakudo >= 0:
            trans = cv2.getRotationMatrix2D((128, 128), Kakudodesu+35-180 , scale)
        else:
            trans = cv2.getRotationMatrix2D((128, 128), Kakudodesu+35 , scale)
        #アフィン変換
        image2 = cv2.warpAffine(shifted_img, trans, (256, 256))

        #主成分軸の面積によって画像を拡大or縮小
        height = image2.shape[0]
        width = image2.shape[1]
        a1 = math.sqrt( 10000/(use_p1_length * use_p2_length) )
        new_img = cv2.resize(image2 , (int(width*a1), int(height*a1)))

        # 保存
        s1 = ''.join(list1[j])
        s2 = s1[-11:]
        #cv2.waitKey()
        #cv2.imshow('src', image2)
        cv2.imwrite('./result_pca/' +onso+ '/' +str(name[num])+ '/syuseibun/' +s2,src) #採択した主成分の図を保存
        cv2.imwrite('./result_pca/' +onso+ '/' +str(name[num])+ '/kaiten/' +s2,new_img) #回転処理をさせた画像を保存
```

**Comment by Anna-Oura** [comment]
# PCAチェックリスト

|音素|check|
|---|---|
| /a/ | ◯
| /i/ | 
| /u/ | 
| /e/ | 
| /o/ | ◯
| /k/ | 
| /g/ |
| /s/ | 
| /z/ |
| /t/ |
| /d/ |
| /n/ |
| /h/ |
| /p/ |
| /b/ |
| /m/ |
| /r/ |
| /w/ |
| /j/ |

**Comment by Anna-Oura** [comment]
- モーラの最初と最後をとりあえずアノテーションしよう. 
  - 最初は動きが大きい. 
  - 最後は唾を飲んだり息を吸ったりする可能性がある. 
  - 頑張ればできそう. 
- 最初は子音からやってみよう. 
- 今は単一音素を発音しているやつだけど, 高速度撮影のやつも見てみたい. 
<br> → 最初は Test2 でやってみよう!!
